hathor/nanocontracts/rng.py

```
@@ -36,17 +46,39 @@ def __init__(self, seed: bytes) -> None:
         cipher = Cipher(algorithm, mode=None)
         self.__encryptor = cipher.encryptor()
 
+    @classmethod
+    def create_with_shell(cls, seed: bytes) -> NanoRNG:
+        """Create a NanoRNG instance wrapped in a lightweight shell subclass.
+
+        This method dynamically creates a subclass of NanoRNG (a "shell" class) and instantiates it. The shell class is
+        useful to prevent sharing classes and objects among different contracts.
+        """
+        class ShellNanoRNG(NanoRNG):
+            pass
+        return ShellNanoRNG(seed=seed)
+
+    def __setattr__(self, name: str, value: Any) -> None:
+        if callable(value):
+            raise AttributeError("Cannot assign methods to this object.")
+        super().__setattr__(name, value)
```

I think this `if` can be more restrictive, I think `__setattr__` is only called (by python) when we set attributes, which are defined in slots. Then you could raise this exception any time `name not in self.__slots__`. You'll probably also have to account for name mangling.

---

I guess I don't have to check whether `name in self.__slots__` because Python does it for me. It will only reach this point if it belongs to `__slots__`. So this code is enforcing that no callables can be assigned (which is true for NanoRNG but it might not be for other classes).

---

> It will only reach this point if it belongs to `__slots__`.

That's not true, `__setattr__` always runs, even if the attribute is not in `__slots__`. What prevents the assign is not Python implicitly, but actually your call to `super().__setattr__`. If you forget that call, the assignment would be allowed.

I guess adding the check doesn't hurt, it's just another protection layer — in the same vein of adding the shell class even though we know class attributes can't be changed because we also used slots there.